### PR TITLE
Add long explanation for E0757

### PR DESCRIPTION
**Pull request by midgleyc**
Helps with #61137

### CODE REVIEW

File: compiler/rustc_error_codes/src/error_codes/E0757.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#![feature(ffi_const, ffi_pure)]
+
+extern "C" {
+    #[ffi_const]
+    #[ffi_pure] // error: `#[ffi_const]` function cannot be `#[ffi_pure]`
+    pub fn square(num: i32) -> i32;
+}
+```
+
+As `const` has a stricter set of requirements than `pure`, remove the `ffi_pure`
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0757.md:15]: `ffi_const` you mean? Also, is there a documentation we could link to by any chance?

Comment (midgleyc) [reply compiler/rustc_error_codes/src/error_codes/E0757.md:15]: Here I mean `const`. `const` and `pure` are C function attributes (e.g. see [Common Function Attributes](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes)). Annotating a function with `ffi_const` says that it's C `const`, and annotating with `ffi_pure` says that it's C `pure`. I could use `ffi_const` and `ffi_pure` instead, but I felt it was more readable to lift the covers up to what they mean in GCC or other C compilers. What do you think?

I could link to [ffi-pure](https://doc.rust-lang.org/nightly/unstable-book/language-features/ffi-pure.html) and [ffi-const](https://doc.rust-lang.org/nightly/unstable-book/language-features/ffi-const.html) from the unstable rust book. The former does mention that `#[ffi_const]` provides stronger guarantees, which implies that it's stricter.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0757.md:15]: Please use `ffi_*` instead then. If I asked the questions, others will wonder about the same thing as well. Please add the links as well then.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup squash

**Pull request merged**